### PR TITLE
adding participant data

### PR DESCRIPTION
**Pull request by marcel-blonk**

### CODE REVIEW

File: lib/jdomodels/src/main/java/org/sagebionetworks/bridge/model/dbo/persistence/DBOParticipantDataDescriptor.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+import org.sagebionetworks.repo.model.dbo.Field;
+import org.sagebionetworks.repo.model.dbo.MigratableDatabaseObject;
+import org.sagebionetworks.repo.model.dbo.Table;
+import org.sagebionetworks.repo.model.dbo.TableMapping;
+import org.sagebionetworks.repo.model.dbo.migration.MigratableTableTranslation;
+import org.sagebionetworks.repo.model.migration.MigrationType;
+import org.sagebionetworks.repo.model.query.jdo.SqlConstants;
+
+/**
+ * Mapping between communities and teams
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/bridge/model/dbo/persistence/DBOParticipantDataDescriptor.java]: please update this comment; "descriptor of what's in a participant data record"

File: lib/jdomodels/src/main/java/org/sagebionetworks/bridge/model/dbo/dao/DBOParticipantDataDAOImpl.java [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+	@Override
+	public void delete(String participantId, String participantDataId) throws DatastoreException, NotFoundException, IOException {
+		MapSqlParameterSource param = new MapSqlParameterSource().addValue(DBOParticipantData.PARTICIPANT_DATA_DESCRIPTOR_ID_FIELD, participantDataId).addValue(
+				DBOParticipantData.PARTICIPANT_ID_FIELD, participantId);
+
+		DBOParticipantData participantData = basicDao.getObjectByPrimaryKey(DBOParticipantData.class, param);
+
+		deleteDataFromBucket(participantData.getS3_bucket(), participantData.getS3_key());
+
+		basicDao.deleteObjectByPrimaryKey(DBOParticipantData.class, param);
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/bridge/model/dbo/dao/DBOParticipantDataDAOImpl.java:173]: always do DB xaction before non-DB xaction to allow rollback

File: lib/jdomodels/src/main/java/org/sagebionetworks/bridge/model/dbo/dao/CsvNullReader.java [hunk 3]
```diff
@@ -0,0 +1,292 @@
+package org.sagebionetworks.bridge.model.dbo.dao;
+
+/**
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/bridge/model/dbo/dao/CsvNullReader.java:3]: please restore test for NullCsvReader

File: client/bridgeJavaClient/src/main/java/org/sagebionetworks/client/BridgeClient.java [hunk 4]
```diff
@@ -126,4 +129,52 @@
...
+	 * @param participantDataId the id of the participantData to append to
+	 * @param data
+	 * @throws SynapseException
+	 */
+	public RowSet appendParticipantData(String participantIdentifier, String participantDataId, RowSet data) throws SynapseException;
+
+	/**
+	 * Upload changed participant provided data
+	 * 
+	 * @param participantIdentifier the de-identified identifier for the participant
```
Comment (brucehoff) [inline client/bridgeJavaClient/src/main/java/org/sagebionetworks/client/BridgeClient.java]: extra line

**Comment by kaysoky** [comment]
Please merge develop back into this pull.  The refactor will break all your new controllers.

To fix, just change all "String username" fields to "Long userId"

File: client/bridgeJavaClient/src/main/java/org/sagebionetworks/client/BridgeClient.java [hunk 5]
```diff
@@ -126,4 +129,52 @@
 	 * @throws SynapseException
 	 */
 	public void removeCommunityAdmin(String communityId, String principalId) throws SynapseException;
+
+	/**
+	 * Upload new participant provided data for this participant for a specific data set
+	 * 
+	 * @param data
+	 * @param participantDataId the id of the participantData to append to
```
Comment (brucehoff) [inline client/bridgeJavaClient/src/main/java/org/sagebionetworks/client/BridgeClient.java]: particpantDataId -> dataDescriptorId

File: integration-test/src/test/java/org/sagebionetworks/IT610BridgeData.java [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+	public void after() throws Exception {
+	}
+
+	
+	@AfterClass
+	public static void afterClass() throws Exception {
+		for (Long id : usersToDelete) {
+			//TODO This delete should not need to be surrounded by a try-catch
+			// This means proper cleanup was not done by the test 
+			try {
```
Comment (brucehoff) [inline integration-test/src/test/java/org/sagebionetworks/IT610BridgeData.java:102]: please don't swallow exceptions

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/DBOBasicDaoImpl.java [hunk 7]
```diff
@@ -189,22 +210,40 @@ public void afterPropertiesSet() throws Exception {
...
+		return getObjectByPrimaryKeyWithUpdateLock(clazz, namedParameters, false);
+	}
+	
+	@Override
+	public <T extends DatabaseObject<T>> T getObjectByPrimaryKeyWithUpdateLock(Class<? extends T> clazz, SqlParameterSource namedParameters)
+			throws DatastoreException, NotFoundException {
+		return getObjectByPrimaryKeyWithUpdateLock(clazz, namedParameters, true);
+	}
+
+	private <T extends DatabaseObject<T>> T getObjectByPrimaryKeyWithUpdateLock(Class<? extends T> clazz, SqlParameterSource namedParameters,
```
Comment (brucehoff) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/DBOBasicDaoImpl.java]: '...withOrWithoutUpdateLock()'

File: lib/jdomodels/src/test/java/org/sagebionetworks/bridge/model/dbo/dao/DBOParticipantDataDAOImplTest.java [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+		DBOParticipantDataDescriptor model = createModel();
+
+		RowSet rowSet1 = createRowSet(new String[] { "a", "b", "c" }, new String[] { null, "1", "2", "3" }, new String[] { null, "4", "5",
+				"6" }, new String[] { null, "7", "8", "9" });
+
+		addToDelete(new DataDeletable(participant.getParticipantId(), model.getId().toString()));
+
+		participantDataDAO.append(participant.getParticipantId().toString(), model.getId().toString(), rowSet1);
+
+		RowSet rowSet2 = createRowSet(new String[] { "a", "b", "c" }, new String[] { "0", "1", "2", "3" },
```
Comment (brucehoff) [inline lib/jdomodels/src/test/java/org/sagebionetworks/bridge/model/dbo/dao/DBOParticipantDataDAOImplTest.java]: please used different data when testing replace

File: services/repository-managers/src/main/java/org/sagebionetworks/bridge/manager/participantdata/ParticipantDataDescriptionManagerImpl.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import org.sagebionetworks.repo.web.NotFoundException;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class ParticipantDataDescriptionManagerImpl implements ParticipantDataDescriptionManager {
+
+	@Autowired
+	private ParticipantDataDescriptorDAO participantDataDescriptorDAO;
+
+	@Autowired
+	private ParticipantDataMappingManager participantDataMappingManager;
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/bridge/manager/participantdata/ParticipantDataDescriptionManagerImpl.java]: ParticipantDataMappingManager:  Please consider renaming to something like ParticipantIDMappingManager

File: services/repository-managers/src/main/java/org/sagebionetworks/bridge/manager/participantdata/ParticipantDataManagerImpl.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+		if (participantId == null) {
+			participantId = participantDataMappingManager.createNewParticipantForUser(userInfo);
+		}
+		return participantDataDAO.append(participantId, participantDataId, data);
+	}
+
+	@Override
+	public RowSet updateData(UserInfo userInfo, String participantDataId, RowSet data) throws DatastoreException, NotFoundException, IOException {
+		List<String> participantIds = participantDataMappingManager.mapSynapseUserToParticipantIds(userInfo);
+		String participantId = participantDataDAO.findParticipantForParticipantData(participantIds, participantDataId);
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/bridge/manager/participantdata/ParticipantDataManagerImpl.java:41]: consider throwing NotFoundException if participantId==null

File: services/repository-managers/src/main/java/org/sagebionetworks/bridge/manager/participantdata/ParticipantDataManagerImpl.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+		return participantDataDAO.update(participantId, participantDataId, data);
+	}
+
+	@Override
+	public PaginatedRowSet getData(UserInfo userInfo, String participantDataId, Integer limit, Integer offset)
+			throws DatastoreException, NotFoundException, IOException {
+		List<String> participantIds = participantDataMappingManager.mapSynapseUserToParticipantIds(userInfo);
+		String participantId = participantDataDAO.findParticipantForParticipantData(participantIds, participantDataId);
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/bridge/manager/participantdata/ParticipantDataManagerImpl.java:52]: NotFoundException if null

**Pull request merged**